src/SAWScript/Crucible/MIR/Builtins.hs

```diff
@@ -1044,6 +1122,41 @@ cryptolTypeOfActual mty =
     baseSizeType Mir.B128  = Just $ Cryptol.tWord $ Cryptol.tNum (128 :: Integer)
     baseSizeType Mir.USize = Just $ Cryptol.tWord $ Cryptol.tNum $ natValue $ knownNat @Mir.SizeBits
 
+-- | Allocate memory for each 'mir_alloc' or 'mir_alloc_mut'.
+doAlloc ::
+     MIRCrucibleContext
+  -> Some MirAllocSpec
+  -> StateT (Crucible.SymGlobalState Sym) IO (Some (MirPointer Sym))
+doAlloc cc (Some ma) =
+  mccWithBackend cc $ \bak ->
+  do let col = cc ^. mccRustModule ^. Mir.rmCS ^. Mir.collection
+     let halloc = cc^.mccHandleAllocator
+     let sym = backendGetSym bak
+     let iTypes = Mir.mirIntrinsicTypes
+     Some tpr <- pure $ Mir.tyToRepr col (ma^.maMirType)
+
+     -- Create an uninitialized `MirVector_PartialVector` of length 1 and
+     -- return a pointer to its element.
+     ref <- liftIO $
+       Mir.newMirRefIO sym halloc (Mir.MirVectorRepr tpr)
+
+     globals <- get
+     globals' <- liftIO $ do
+       one <- W4.bvLit sym W4.knownRepr $ BV.mkBV W4.knownRepr 1
+       vec <- Mir.mirVector_uninitIO bak one
+       Mir.writeMirRefIO bak globals iTypes ref vec
+     put globals'
+
+     ptr <- liftIO $ do
+       zero <- W4.bvLit sym W4.knownRepr $ BV.mkBV W4.knownRepr 0
+       Mir.subindexMirRefIO bak iTypes tpr ref zero
+     pure $ Some MirPointer
+       { _mpType = tpr
+       , _mpMutbl = ma^.maMutbl
+       , _mpMirType = ma^.maMirType
+       , _mpRef = ptr
+       }
```

Comment:
This is where most of the magic powering `mir_alloc{,_mut}` happens. Using the `crucible-mir` memory model, this allocates a fresh reference, writes uninitialized memory to it, and returns the resulting pointer. This code was largely inspired by [similar code in `crucible-mir`](https://github.com/GaloisInc/crucible/blob/c19fc2c53851cdd162ae1ccca14032467e3c2fd1/crucible-mir/src/Mir/TransCustom.hs#L1393-L1402).